### ISSUE TITLE
Cockatrice on Android / iOS / mobile / phone / tablet

### ISSUE THREAD

**Issue by isleep2late** [Cockatrice/Cockatrice#646]
I spoke with some people, including ZeldaZach and Daenyth, about getting Cockatrice to work on mobile. Originally I wanted to be able to open the chat on my iPhone and just ask people questions/talk, but it appears getting all aspects of Cockatrice onto iOS would be a more high demand project. I'll leave here a preliminary to-do list for the hypothetical case in which some of us would be interested to pick up this project, along with some follow-up points/ideas/concerns.
1. GUI - Some sort of user interface/framework would have to be fashioned, and this first step I'm guessing will be the hardest to get through because it requires some original code.
2. Transcribe the original C++ code to compile on Xcode - At least one source claims Xcode (the program used to develop/compile iOS apps for Macintosh), accepts not only Objective-C but most other C-oriented languages including C++. I've tried copy-pasting some of your Cockatrice source code onto the relevant portions of where it would belong to an iOS app, and surprisingly most of it was readable, with the exception of minor syntax/the obvious unrecognized objects/variables. My hope is that this is all true, and once a GUI is made all that is required is to tweak the C++ code to an acceptable format for Xcode.
3. Compiling/Debugging/Proofreading/Editing - I really don't see this as a one-man job, and this step is really crucial. We'd need multiple people with access to an iPhone 5/6/6+ and iPad to confirm that this will work on their device. We need someone else other than me to have Xcode on a Mac or some means by which to compile the code on an iOS emulator. We'll need other members of the community who'd be willing to help look over any diagnostic issues. Which segues to my next point...
4. Seek help for any problems from #1-3: This part of the checklist is also my opportunity to stress the need for this to be a teamwork effort. I believe in the past people have fantasized over this idea individually without any collective effort to act on it. I'm not saying this will happen, or that it should happen, but that IF this would ever happen in the future, these are some of the steps that we would need to take.

Follow-up concerns:
- Clearly this would get wrecked from a legal standpoint, so I would recommend the app be developed for Cydia rather than the App Store. There's tons of apps that make it to iOS via Cydia with variable success that would otherwise not be publishable to the App Store for logistical reasons. This would severely limit the scope of audience however...
- Perhaps publish to App Store at an obviously free cost without advertisements or profit (I'd recommend the same even if this went to Cydia). Advertise this as a "Sandbox" program [which it is] independent of all ties to MtG, but enable users access to certain card databases for download, such as Magic, or any other card game.
- I don't know how high the demand for this would be, so I guess I'll just leave this here and assume the demand is proportional to how much contribution there is to the project.
- This would be a free, open-source project so it really would take a lot of time if this ever happens. I go to school with a bunch of other priorities I need to worry about and I'm sure the rest of us have very busy lives. At most I can probably delegate, contribute a couple hours on weekends to code, and offer design concepts or ideas for improvement.
- These are all ideas that one person is just throwing out there, so it's perfectly fine if this ends up not amounting to anything, but some people have asked for something like this in the past and it wouldn't hurt to at least gather some interest for developers. So please, take all of this for a grain of salt if you don't see this happening.

**Comment by ctrlaltca** [comment]
I currently have a private branch with cockatrice running on android, and easily portable to ios.
I said "running" and not "working" because there are a lot of UI problems that needs to be solved, but the code is running fine.
I currently wanted to keep the branch private to avoid getting under the eye of sauron..

**Comment by Daenyth** [comment]
Honestly I think the best bet is to start from scratch with the .proto files and make a fully native UI.

**Comment by isleep2late** [comment]
@ctrlaltca thanks for the heads up! Glad to know that people have worked on something like this before, and to know that it is in fact possible. I am more or less mainly interested in an iOS build, but let us know if you'd ever feel comfortable making your private android branch public.

@Daenyth Starting from scratch is also a possibility, and would probably yield a better finish. It would be harder though, so I'm hoping that there are people who use cockatrice who happen to be familiar with iOS development who are looking at this post

**Comment by ZeldaZach** [comment]
You can always ask around in the main chat @isleep2late

**Comment by Kayoscape** [comment]
I don't know if this project is still happening, but if it is then I'd love to be a part of it. I have object oriented programming experience, but I haven't used Xcode before, nor have I written anything in Objective-C. Still, I hope I can be of some help. Please let me know what I can do.

**Comment by Daenyth** [comment]
I think the only way to do this sanely would be to start from scratch. Take
the proto files and make a native ui. The client itself doesn't have much
logic in it, mostly just ui and how to update visuals based on game actions

**Comment by LordKahra** [comment]
Hello, I'm an Android developer and while I was learning to code I did a lot of work on a playtesting app. It's not finished, and it's very rough due to my newness to code while I was working on it, but I'd be willing to go through it and see if a rough draft could be made to work with Cockatrice. I loosely followed MVC, so if nothing else I think a workable (if ugly) UI could be pulled from it.

I've never really contributed to an open source project before, is there someone I should consult before making a fork/adding a folder for Android related development/etc?

**Comment by Daenyth** [comment]
It would be best to make a new github project and load in the protobuf
files via git submodule

**Comment by alexchandel** [comment]
1. You don't need to make this a Cydia app. Apple no longer prohibits side-loading apps: you can download, compile, and install third party apps for free. There are no legal consequences to this.

2. Xcode is just an IDE. It's a front-end to the compilers, `clang` and `swift`, which can compile C, C++, Objective-C, Objective-C++, OpenMP, OpenCL, CUDA, and other C-derived languages, and of course Swift.

3. If your deck management/server communication/game-finding/game state code is all already in C++, and your networking library supports iOS, then I'd recommend keeping it. If there's truly nothing on the client though, and you can generate Swift implementations of these protocols, then use Swift.

4. The UI should be written specifically for iOS, but this shouldn't be too hard. Objective-C is under-appreciated—far better than Java/C#/other nightmares—but since you're writing a brand new UI I'd recommend using Swift.

**Comment by ClintFrancis** [comment]
This could be ported for both iOS and Android using Xamarin.
The UI could be done using CocosSharp or SkiaSharp

**Comment by Daenyth** [comment]
It's open source, go for it.

**Comment by alexchandel** [comment]
Xamarin's Cocoa support is meager unless you're willing to pay for the developer's license. And building Cocoa interfaces programmatically is a pain compared to laying them out on XCode. And the UI code would have to be specific to iOS anyway, since CocosSharp doesn't port to Android.

It's (substantially) easier to make the iOS UI in Xcode, which has a GUI layout tool for iOS and can compile C and C++ code.

**Comment by ClintFrancis** [comment]
Full disclaimer: I work for Xamarin and have been developing using the tools for five years. 

Xamarin has full support for using layouts created in XCode. You can either edit / create them directly in XCode or within Xamarin studio. We also offer the same workflow with Android using Android Studio. 

CocosSharp has full support for both Android and IOS. You can check out some of the samples here. 
https://developer.xamarin.com/samples/mobile/BouncingGame/

As for the price of tooling Xamarin wad made free for community development and also available as part of an msdn subscription.

**Comment by Daenyth** [comment]
I feel like this is getting a bit off topic. Let's focus on what we can for
Cockatrice on mobile

**Comment by alexchandel** [comment]
@ClintFrancis support for nib files isn't the same as storyboard integration. And I'm not saying Xamarin Studio isn't free, but last time I checked the newest iOS and OSX cocoa bindings required payment. Has this changed?

But in any case I'd recommend native code. There's a speed, memory, and power usage advantage, and Swift is an excellent language. The Android team is currently considering making it a first class language on Android now that it's open-sourced.

@Daenyth Is the deck/network/game logic of Cockatrice relatively disjoint from the GUI? As in the Model-View-Controller pattern? Because if so, it should be possible to write a very thin GUI wrapper, like a `.storyboard` and single `.swift` file, which are controlled by the existing game model & game controller logic.

**Comment by ctrlaltca** [comment]
I'd use Emacs. Oh sorry, i meant to say Vi. 🔥 
Please have a look at what's in the code before arguing what's easier and what's better for cockatrice.

I already got trice running on Android, and since the code is based on Qt, the obvious choice was Qt creator; it was just a matter of fixing the build files to get it running.
The desktop UI is obviously far from being usable on a phone, i had a few games on a tablet but imho the "right-click-menu" and "keyboard shortcut" based interface of cockatrice can't be fixed for a touch interface. 
Maybe a QML / Qt quick interface would be the easiest to glue with the current code.

> Is the deck/network/game logic of Cockatrice relatively disjoint from the GUI?

Most of the low-level stuff (decks, game zones, etc) is shared between the client and the server, so it has no dependencies on the UI widgets.
All the IO (both file and networking) stuff has dependencies on Qt libraries, most of the data structs are Qt based.

> As in the Model-View-Controller pattern?

Qt itself uses MVC, but their implementation is so tight to their own classes that i can't really think of easily plug them to something else.

> Because if so, it should be possible to write a very thin GUI wrapper, like a .storyboard and single .swift file, which are controlled by the existing game model & game controller logic.

Developing a solution specific for apple devices would cut out android users.

At the end of my experiments with porting the code to mobile, i'd changed my mind and started creating a webapp. The app is currently able to connect, lets you join rooms and chat with users.
Implementing all the game-related procedures is still a long task anyway..

**Comment by alexchandel** [comment]
@ctrlaltca 

> Developing a solution specific for apple devices would cut out android users.

The point is that there may exist a simple, specific solution for each.

> Most of the low-level stuff (decks, game zones, etc) is shared between the client and the server, so it has no dependencies on the UI widgets.
> All the IO (both file and networking) stuff has dependencies on Qt libraries, most of the data structs are Qt based.

They look very reusable.

> Qt itself uses MVC, but their implementation is so tight to their own classes that i can't really think of easily plug them to something else.

That's reassuring enough, since Qt provides mobile widgets.

> The desktop UI is obviously far from being usable on a phone, i had a few games on a tablet but imho the "right-click-menu" and "keyboard shortcut" based interface of cockatrice can't be fixed for a touch interface. 
> Maybe a QML / Qt quick interface would be the easiest to glue with the current code.

I'd imagine keyboard shortcuts have no replacement without an external keyboard, but the "right-click-menu" interface can be replaced with mobile UI pattern like "long-press"->"modal-dropdown". In the end it may come down to a separate, slightly modified mobile GUI.

**Comment by Daenyth** [comment]
@alexchandel long press for dropdown would make a UX nightmare. The majority of game actions period are right click or keyboard based. This is not tenable as a solution.

I'm quite honestly getting pretty tired of this issue. It's come up repeatedly over the past 5-6 years, and every time it does, I've asked for people to show some kind of UI concept that isn't painful; mockups, even a text description of how a game would play out with a touch based interface. Not a single person has replied to any of these, and as you can see, one of the few people who has actually *tried* to port the codebase over has also not been able to see a path to making a usable touch UI.

We'd all be extremely happy to accept PRs adding any necessary foundation for this. We'd be happy to provide assistance, guidance on how the codebase works, or the protocol, etc.

What I'm damn tired of is more bikeshedding about how to hypothetically do it. I'd like the rest of the conversation on this ticket to cover:

- Discussion of UX workflows
- Mockups
- Specific technical discussion of specific code changes needed to support native porting
- Links to actual attempts to do the porting, or specific blockers that need to be resolved to support such an attempt.

In the interest of not wasting everyone's time, I'll have little choice but to lock the discussion on this if it doesn't get productive.

**Comment by Daenyth** [comment]
To get the ball rolling on that more productive conversation;

Every app I've seen on mobile for card games has the main interaction being context-sensitive drag actions. This requires a rules-aware game engine. Given that cockatrice's goal is to be a rules-unaware sandbox, this poses a problem.

I also think there's a problem with information density. Compare the board state at a glance between Ascension (works really really well on mobile) vs MtG. Consider the number of cards. Attributes per player. Imagine a 4-person EDH game shown on a 5" screen. Visible enough that you can realistically play against someone on a computer and be able to interject in the chat when you need to respond instead of passing priority. These are the challenges I'm talking about. Not framework, IDE, or compiler. Those are small problems and easily solved. UX is always the hardest part.

**Comment by Kayoscape** [comment]
Not sure if this is helpful, but I have a few UX ideas.   
  

  
Drag actions do not necessarily need to be context-sensitive. Dragging in Cockatrice merely represents moving a card, which can be done with any card at any time to any destination on the desktop app, so I don't see a reason to change that behavior. Dragging can be triggered by holding a card for at least one (two?) seconds.   
  

  
As for the space issue, I propose that the user can only see one player's side of the battlefield at a time, and can switch to a different one using buttons on the sides of the screen. Cards can also be small (take advantage of high resolution on mobile) and perhaps don't need all information at a glance. A tap on a card can open a detailed view with a context menu for actions. I also recommend that cards be auto-alligned into a reorder-able grid to prevent issues with movement on small screens.   
  
  
  

  
Thoughts?
  

  
>

**Comment by skwerlman** [comment]
Instead of using real estate for buttons to switch sides of the board, you could instead use a swiping motion to slide between different players. An upward swipe would return you to your side, and a downward swipe would open a context menu for the currently selected card(s).

**Comment by Daenyth** [comment]
@Kayoscape I think some kind of reduced view will be necessary, but we need to make sure that you can see everything going on.

There are a few different workflows which may have different needs.

- Playing 1v1
- Spec 1v1
- Playing MP
- Spec MP

**Comment by spooky-san** [comment]
As much as I don't like taking away control from the user, if we decide to stick to having only the user's side of the board on the screen it could switch to the opponent's board if anything on their side updates (i.e. tapping a card, playing a card).

 If we decide that maybe having both sides on the board on the screen at the same time is the better path to take, then perhaps making the opponent's board smaller than the user's board is a more efficient use of space.

Building upon @skwerlman 's idea, I think perhaps we have 3 different views:

1. both sides of the board and the user's hand

2. user's side of the board and hand

3. opponent's side of the board

Additionally, the space on the phone is rather limited, but considering the amount of vertical space required to show all of the information, perhaps a vertically oriented UX might be more intuitive here than a horizontally oriented one?

Alternatively, we could use both and design each view around one or the other.

Consider the first view (which might be our default view), it could be designed with a more vertical orientation in mind, allowing more board to be displayed on the screen, but less detail. If such is the case, then it may be a good idea to design views 2 and 3 with a horizontal orientation so that we may show more detail on either side of the board.

In terms of accessing these views, I think that if the user has their phone oriented vertically, it'll remain in view 1.  If the user has their phone oriented horizontally, they can switch between view 2 and 3 by swiping down and up respectively.

**Comment by Daenyth** [comment]
What do you think about explicitly targeting tablet size screens first
before trying to fit it all in a phone size? It would let us get some feel
for what works and what doesn't

**Comment by spooky-san** [comment]
Oh, yeah we could definitely target tablet sizes. However, I think we'd have to develop different UX's for each device due to screen sizes.

When it comes to the tablet, I think we could try maybe 3 views, all horizontal. We'll try all three of the views I described above being horizonal, and see how that works. Then, we try a vertical view for the tablet and see if that feel's better than the horizontal on view 1. If it is then we keep it and see where we move from there. However, if we find that neither feels okay, we'll have to try to design a new default view.

When it comes to the tablet:

For the menu(i.e. shuffling a deck, showing a hand...) I think we should tap and hold on blank space, This will bring up a contextual menu, similar to a dropdown menu, but instead it's just a menu that you switch to when you hold your finger on something(i.e.  reveal hand if you hold on the box where you hand,...), similarly to the way a drop up menu shows up when you right click someway.

Regardless of what happens, mobile controls will always be (at the very least), slower than the controls on desktop.

**Comment by DrKittens** [comment]
@ctrlaltca

@Daenyth 

>I've asked for people to show some kind of UI concept that isn't painful; mockups, even a text description of how a game would play out with a touch based interface. Not a single person has replied to any of these, and as you can see, one of the few people who has actually tried to port the codebase over has also not been able to see a path to making a usable touch UI.

Ideas for the GUI for a TWO PERSON GAME!!!

1. Phone orientation - We force landscape
2. Hand location - Bottom of the screen
3. Small card size, 3 rows above hand = field, colour code them
Front row (creatures and shit) = red
Middle row (enchants and other shit = blue
Backrow (resources) = yellow
4. Ingame chat/log accessible via a tab, viewing card names functions by long pressing on the cards same as desktop.
5. The stack - Whenever a spell is played a beep is made and the card is popped up on the screen (since cockatrice isnt "smart" this would include lands) this card appears fullscreen for X seconds and after that expires the card hits the field and play continues. If the opponant wishes to read the card OR respond with a counter they can click a "wait" button under the card and then play something from hand or click a "nvm lol" button, if no card is played by the opponant after XX seconds the game continues, if an opponant plays a card then this repeats til no-one plays a card then it lists the cards left-to-right last-to-first (aka how the stack fucking resolves)

6. There are 3 views toggleable by a button in a corner
1-My board
2-My opponants board
3-The front row of BOTH our boards + an icon that represents the opponant/the player.

6. The game steps (draw combat etc) are shown on the TOP of the screen, to draw a card DRAG the draw step to your hand (top of phone to bottom of phone) (can be done on any phase).

7. Advancing to battle/combat phase changes the view automagically to view 3 (both players front rows)

8. Pressing the "end turn" button step makes the opponants phone beep, they click on their "untap" button to untap and take the turn. Both players can still do stuff but the person who clicked "end turn" can no longer change the step back to something else.

THIS IS NICE AND ALL BUT HOW DO I PLAY GAEM??

To view card info: Long click on ANY card opens it full screen, tapping anywher eon the screen leaves this view
To "tap" a card: short single tap
To "move" a card: Cards can only move between rows, drag the card to a different row.
To "point" cards to other cards drag a card to another card/player icon.
To "add counters/make related card/SET power toughness/send to grave/duplicate card/send to exile/SET counters" double tap a card to open a context menu. **set not add, we want an input option for numbers here for speed...**

To "select multiple cards" button next to hand to get a "selection" tool. Single tap cards to select them the press the tool button again to confirm selection. All actions made (drag/double tap context menu etc) made to any of the "grouped" cards will be made to all of them.

Possibly allow users to make multiple groups of multiple cards, ie all creatures in 1 group, all lands in another etc...

DECK/GY/EXILE piles are shown somewhere (top of screen after phases!?)
To perform actions with them double tap the piles to open relevant context menus.
View/shuffle/scry/search etc...

Deck view = fullscreen, provide a search bar or make multiples of cards stack with like 4x in the corner.
autosort by card type and shuffle on close by default checked..
This will be cluttered as fuck tbh and it may be worth showing it portrait but thats probably bad design since the rest of the game is landscape?

Am i missing anything else!?

**Comment by Daenyth** [comment]
@DrKittens I recommend drawing it. Start with your text description and just mock out what a game state could look like. Maybe on paper with one-to-one sizes.

**Comment by DrKittens** [comment]
Detailed how a short game would play out, i'd draw it but its like 3am, probably come back to it later/when i dont have work in 6h lmao..

**Comment by arloft** [comment]
I know this is an old discussion, but I just saw it. I’m a UX / UI designer and would love to help out porting the visual design to mobile. I think this is a tough nut to crack but doable if we are willing to approach mobile as a legit branch of the app, as so much of the UI and UX would need redesigned. I dabble in code and actively work with a developer on my professional gig, so I’m accustomed to developer speak as well as prepping visual assets for the dev team. 

Regarding installation to iOS devices, the App Store is out for obvious reasons. Also, I feel Cydia isn’t the best way either as it’s too reliant on a now unstable and irregular jailbreak pattern. However, I would suggest taking the approach used by GBA4iOS, using an enterprise install model that gets around the App Store.
http://www.gba4iosapp.com/download/
The source is linked on their site (resides on butbucket).

Long story short I guess the question I have is, are efforts toward a mobile port still active?

**Comment by Daenyth** [comment]
No. Mockups for some of the main interactions are a big next step

**Comment by arloft** [comment]
Is there any existing documentation or a collection of thoughts / proposals
about how to approach the limited screen size of a phone? I ask only so I
don’t needlessly retread anything that’s already been mentioned.

**Comment by Daenyth** [comment]
This page is about it

**Comment by LizzieontheBay** [comment]
All I know is I’m desperately in need of help!! There is no way I can learn all of this. These are more than hackers these criminals are pro at what they are doing and they have me in fear all the time! It’s obvious someone is paying them and the only person I have had a problem with is my ex husband who is a doctor and my daughter who I disowned who is a heroin addict but very savvy on computers and the dark net and has friends who are the same. Please have someone who has been working on my problems contact me tomorrow as soon as they can. These criminals have harmed me physically too!! Even sabatoging my car with fumes and numerous other things that could have killed me. I cannot contact FBI from any device I have and since their equipment is remote and they even have devices that are remote that they take to the road with following me and get into every internet I try to use anywhere to get the FBI involved. They have even broken the law crossing state lines to criminally hack me and destroy property. Please notify FBI ASAP PLEASE! My number is+15044503989. I pray to Dear God you can do this for me and also contact me tomorrow as soon as you can! My life is in danger!!!!!!! THANK YOU SO SO MUCH!!!!! Lizzie ( Elizabeth L. Giambelluca)

**Comment by LizzieontheBay** [comment]
PLEASE HELP!!!!!!!!!!!!! Lizzie

**Comment by Nightfirecat** [comment]
@LizzieontheBay In case you're being serious, I would advise you go see your local police department in person. What you're describing is not something that random strangers on the internet will be able to help you with.

**Comment by Daenyth** [comment]
troll comments removed...

**Comment by LizzieontheBay** [comment]
The local police doesn’t do crap!!!!

**Comment by LizzieontheBay** [comment]
I’m dead serious and they started doing things to cause harm to me and even
was pumping some kind of chemical like carbon dioxide and nitros oxide and
so mush Freon that my puppy and I our nose was bleeding and my skin was
blistered and we got choked and couldn’t breathe. My ex is a parish
official and he already has been in trouble. If you dont contact FBI to
help we could be in serious danger

**Comment by arloft** [comment]
I’m still working on this, the main interactions. Trice doesn’t lend itself easily to a phone sized, no cursor, no hover kind of environment. Think I’m close to a basic gameplay field that works on a phone, hoping to put up some wireframes soon. What’s the preferred way to share those with developers? I can put them on google drive and link to the folder, allowing anyone to comment? Attaching them to this discussion seems like it would be tough to link discussion to their relative wires/mockups.

**Comment by Daenyth** [comment]
That sounds great! Drive is fine

**Comment by papb** [comment]
Is there anything I can help with? Ideally something that don't really need a lot of architecural understanding of cockatrice, some specific module?

**Comment by Rendroc** [comment]
Has any progress been made on this? I’d really like to be able to play on my iPad.

**Comment by ebbit1q** [comment]
Progress on this has been frozen for a year, this seems to be the only thing that was made: https://github.com/arloft/Mobitrice--cockatrice-mobile
Of course you can help, this is open source software after all, what we need is a port of the interface.

**Comment by arloft** [comment]
The github page linked above is mine, and it's mainly a UXUI concept (I
should add the concept images into this repo, I'm doing the design on
Figma.com) for how Cockatrice might function on a phone. It basically
involves new apps for iOS and Android, and a possible re-think of how the
server communicates game state to the client, but I'm trying to not suggest
changes to the server operation, and design solely within the confines of
the client app itself.
I have dropped the ball on furthering this design as I had some changes in
my personal which are now mostly settled. I should be able to pick this
project back up and get more design work to the dev team in the near future.

**Comment by tooomm** [comment]
For Cockatrice on Android check
https://github.com/Cockatrice/Cockatrice/wiki/Cockatrice-on-Android-(phone-tablet)

(Taken from https://github.com/Cockatrice/Cockatrice/issues/3899#issuecomment-601082273, thanks @vindicate7)

**Comment by ebbit1q** [comment]
Do note that this article is mainly a proof of concept rather than a recommendation.

**Comment by ctrlaltca** [comment]
Closing this issue. While it's been a nice discussion, it's really unrelated to the current development of Cockatrice.
If someone wants to work on this, we're willing to setup a new repository under the umbrella of the Cockatrice project and move all the discussion in there.
